packages/tpetra/core/test/MultiVector/MultiVector_MicroBenchmark.cpp

```diff
@@ -0,0 +1,149 @@
+//
+// ************************************************************************
+// @HEADER
+*/
+
+#include "Tpetra_TestingUtilities.hpp"
+#include "Tpetra_Map.hpp"
+#include "Tpetra_MultiVector.hpp"
+#include "Teuchos_ScalarTraits.hpp"
+#include "Kokkos_ArithTraits.hpp"
+#include "Tpetra_Details_Profiling.hpp"
+#include "KokkosBlas.hpp"
+
+namespace { // (anonymous)
+
+  using Tpetra::TestingUtilities::getDefaultComm;
+  using Teuchos::RCP;
+  using std::endl;
+
+  //
+  // UNIT TESTS
+  //
+
+  ////
+  TEUCHOS_UNIT_TEST_TEMPLATE_4_DECL( MultiVector, MicroBenchmark_Update, SC, LO, GO, NT )
+  {
+    typedef Tpetra::Map<LO, GO, NT> map_type;
+    typedef Tpetra::MultiVector<SC, LO, GO, NT> MV;
+    typedef Teuchos::ScalarTraits<SC> STS;
+    
+    SC ONE = STS::one();
+    SC ZERO = STS::zero();
+
+    // Problem parameters
+    size_t vector_size = 125000;
+    size_t num_repeats = 2000;
+
+    auto comm = getDefaultComm ();
+    const int numProcs = comm->getSize ();
+    if(numProcs != 1) return;
+
+    const LO lclNumRows = vector_size;
+    const GO indexBase = 0;
+    RCP<const map_type> map (new map_type (lclNumRows, lclNumRows,
+                                           indexBase, comm));
+    const LO numVecs = 1;  // FIXME: This needs to be 1 for the lambda loop to work
+    MV X (map, numVecs), Y(map,numVecs);
+    X.putScalar(ZERO);
+    Y.putScalar(ONE);
+
+    Kokkos::fence();
+
+
+    typedef typename MV::dual_view_type::t_dev dev_view_type;
+    //    typedef typename MV::dual_view_type::t_host host_view_type;
+    typedef typename dev_view_type::memory_space cur_memory_space;
+    typedef typename dev_view_type::execution_space cur_exec_space;
+
+    // Run Tpetra Update Loop
+    {
+      ::Tpetra::Details::ProfilingRegion region ("Tpetra Update Loop");
+      for(size_t i=0; i<num_repeats; i++) {
+        X.update(ONE,Y,ONE);
+      }
+    }
+
+    X.putScalar(ZERO);
+    auto X_lcl = X.template getLocalView<cur_memory_space> ();
+    auto Y_lcl = Y.template getLocalView<cur_memory_space> ();
+    Kokkos::fence();
+
+    // Run KokkosBlas Update Loop
+    {
+      ::Tpetra::Details::ProfilingRegion region ("KokkosBlas Update Loop");
+      for(size_t i=0; i<num_repeats; i++) {
+        KokkosBlas::axpby(ONE, X_lcl, ONE, Y_lcl);
+      }
+    }
+
+    X.putScalar(ZERO);
+    Kokkos::fence();
+    Kokkos::RangePolicy<cur_exec_space> policy (0, vector_size);
+    // Run raw lambda loop
+    {
+      ::Tpetra::Details::ProfilingRegion region ("Raw Lambda Update Loop");
+      for(size_t i=0; i<num_repeats; i++) {
+        Kokkos::parallel_for(policy,KOKKOS_LAMBDA(const size_t &i) {
+            // This is what we should do in general
+            //            X_lcl(i,0) = ONE * X_lcl(i,0) + ONE * Y_lcl(i,0);
+            // This is the special case code that actually executes inside KokkosKernels for alpha=1
+            X_lcl(i,0) += ONE * Y_lcl(i,0);
+          });
+      }
```

Comment:
You actually want the fence here, within the timing scope but outside (and after) all the repeats.  This will ensure accurate timings, at least if we can ever turn off `CUDA_LAUNCH_BLOCKING`.